Choose how to justify circuit with latex drawer

maddy-tod opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
This implements the justify parameter as described in #1725 for the latex drawer.

The change consists of now iterating over the gates, so that they aren't just drawn in their own layer anymore.

mtreinish commented on CHANGELOG.rst:
```diff
@@ -27,8 +27,10 @@ Added
 - Added support for register slicing when applying operations to a register (#1643).
-- Added in new parameter ``justify`` to the text circuit drawer to say how the
+- Added in new parameter ``justify`` to the text, mpl and latex circuit drawers to say how the
   circuit should be aligned. (#1725)
```
I'd probably add the 3 PRs here for each backend since 1725 didn't add it to all the backends.

maddy-tod replied:
I have done

mtreinish commented on qiskit/tools/visualization/_circuit_visualization.py:
```diff
@@ -459,4 +467,5 @@ def _matplotlib_circuit_drawer(circuit,
                                        reverse_bits=reverse_bits)
-    return qcd.draw(filename)
+    qcd.parse_circuit(circuit)
+    return qcd.draw(filename, verbose=True)
```
Why is this being changed here? We should keep this PR to just modifying the latex path, it's complicated enough without bundling mpl changes in it.

maddy-tod replied:
This shouldn't be here - the `parse_circuit()` function was merged into the `draw()` one in the mpl PR

maddy-tod commented:
I have removed the `_get_instructions()` function from `_utils` and updated the tests that were using it to now use the `_get_layered_instructions()` function. However, in doing this I noticed that sometimes layers can be returned blank (hence the presence of `[]` in the testing). It is trivial to fix this, however it does mean some of the test of the text drawer need to be updated as they will now contain slightly less blank space, so I will do this as a separate PR.

maddy-tod commented:
I think that #2039 is needed to make this PR pass the test `test_normal_circuit` consistently, as this is a randomly generated test. It passes ~80% of the time currently but #2039 should make it 100%.

mtreinish commented on qiskit/visualization/latex.py:
```diff
@@ -222,236 +222,243 @@ def _get_image_depth(self, aliases=None):
+
+        boxed_gates = ['u0', 'u1', 'u2', 'u3', 'x', 'y', 'z', 'h', 's',
+                       'sdg', 't', 'tdg', 'rx', 'ry', 'rz', 'ch', 'cy',
+                       'crz', 'cu3']
```
Why did you remove 'id' from the list?

maddy-tod replied:
I kept an older version of the list, I have added `'id'` back in.

mtreinish commented on qiskit/visualization/matplotlib.py:
```diff
@@ -520,7 +520,7 @@ def _draw_ops(self, verbose=False):
 
                 if verbose:
-                    print(op)
+                    print(op.name)
```
This shouldn't be changed in a commit for latex

maddy-tod replied:
Removed

mtreinish commented on qiskit/visualization/latex.py:
```diff
@@ -508,7 +515,6 @@ def _build_latex_array(self, aliases=None):
         We will deduce the register names and sizes from aliases.
         """
         columns = 1
-        is_occupied = [False] * self.img_width
```
Is the reason this being remove (and all the associated logic below) because we rely on the layers to determine when we need to shift because of a barrier? If so this will be useful because tracking this was a real pain and a constant source of bugs.

maddy-tod replied:
Yes the layers say which column things should be drawn in, so there is no need to keep track of which spaces are occupied anymore.

mtreinish commented:
> This is now passing the testing, however there is an issue with drawing barriers. For example, the following code results in an error
> 
> ```
> q = QuantumRegister(3, 'q')
> qc = QuantumCircuit(q)
> 
> qc.h(q[0])
> qc.barrier(q[1], q[2])
> 
> circuit_drawer(qc, filename='test_latex.png', output='latex')
> ```
> 
> `Undefined control sequence. <recently read> \barrier `
> I also get this message when I try to run the code against master, so I think this may be a separate issue.

I'm not able to reproduce this on master. When I run the script I get this output image (which looks weird and is typical of the spacing bugs on latex with barriers):

![test_latex](https://user-images.githubusercontent.com/2447371/55829264-447f2900-5adc-11e9-86da-94299e0bca2e.png)

I believe that error you're hitting is because you don't have the latest version of the qcircuit package from CTAN. We added `\barrier` in the latest version, 2.6.0

maddy-tod commented on qiskit/visualization/circuit_visualization.py:
```diff
@@ -459,4 +469,4 @@ def _matplotlib_circuit_drawer(circuit,
                                        plot_barriers=plot_barriers,
                                        reverse_bits=reverse_bits)
-    return qcd.draw(filename)
+    return qcd.draw(filename, verbose=True)
```
Shouldn't be, removed.

maddy-tod commented:
@mtreinish you were exactly right, thank you! I think it would be good to update the error thrown to tell users that this may be the issue they were having, but that should be in a separate PR.

mtreinish commented:
So this still LGTM, but doing some final local testing before approving I found a weird quirk when setting justify='right'. There are a lot of blank wires being appended to the end of the circuit making the image unnecessarily large. For example, with a random test script I had lying around (from testing an earlier latex bug):

```
q_a = QuantumRegister(3, name='qa')
q_b = QuantumRegister(5, name='qb')
c_a = ClassicalRegister(3)
c_b = ClassicalRegister(5)

circuit = QuantumCircuit(q_a, q_b, c_a, c_b)

circuit.x(q_a[1])
circuit.x(q_b[1])
circuit.x(q_b[2])
circuit.x(q_b[4])
circuit.barrier()
circuit.h(q_a)
circuit.barrier(q_a)
circuit.h(q_b)
circuit.cswap(q_b[0], q_b[1], q_b[2])
circuit.cswap(q_b[2], q_b[3], q_b[4])
circuit.cswap(q_b[3], q_b[4], q_b[0])
circuit.barrier(q_b)
circuit.measure(q_a, c_a)
circuit.measure(q_b, c_b)
print(circuit.draw(output='latex_source'))
circuit.draw(output='latex', filename='/tmp/circuit_work.png', justify='right')
```

Returns:

![circuit_work](https://user-images.githubusercontent.com/2447371/55880405-9f616080-5b6e-11e9-8e99-358a23e64ec4.png)

but when I ran that without a justify parameter (so left justified) like:

`circuit.draw(output='latex', filename='/tmp/circuit_work.png')` it returned:

![circuit_work-bak](https://user-images.githubusercontent.com/2447371/55880450-b3a55d80-5b6e-11e9-873f-695cdcde39e3.png)

I also tested it with mpl right justified just to compare:

`circuit.draw(output='mpl', filename='/tmp/circuit_work-mpl.png', justify='right')`

![circuit_work-mpl](https://user-images.githubusercontent.com/2447371/55880615-1696f480-5b6f-11e9-9680-512093716b42.png)

I'm still tracing through the code to try and figure out what's causing this, but we shouldn't be appending all those blank wires to the end if they're is no element there.

mtreinish commented:
Err, actually just found another issue using the same test circuit from above. Trying it with `justify='none'`  the same circuit fails on an index error:

```

Traceback (most recent call last):
  File "/home/mtreinish/git/qiskit/qiskit-core/qiskit/visualization/latex.py", line 901, in _build_latex_array
    self._latex[pos_1][column] = "\\meter"
IndexError: list assignment index out of range
```

maddy-tod commented:
> There are a lot of blank wires being appended to the end of the circuit making the image unnecessarily large
I think this is due to the depth of the circuit being incorrectly calculated, I will try and push a fix soon.

maddy-tod commented:
I have pushed a fix which seems to fix both of these problems. It is quite a large change, but has worked on all the examples I have tried so far. 

There is still an issue that the `None` parameter doesn't work quite as expected - the gates aren't necessarily placed in the order they were input into the circuit, but they are all in individual columns. This is due to the fact `dag.nodes_in_topological_order()` sorts by qargs. I think this could be fixed, but it is a separate issue and affects all the drawers.

mtreinish commented:
> I have pushed a fix which seems to fix both of these problems. It is quite a large change, but has worked on all the examples I have tried so far.

Nice, that's great! I'm glad to see that convoluted depth function reworked to be much simpler. It's been a constant pain keeping that updated and working in the past. Relying on the layers makes a lot more sense. I confirmed this fixes the 2 issues identified earlier. 

> There is still an issue that the `None` parameter doesn't work quite as expected - the gates aren't necessarily placed in the order they were input into the circuit, but they are all in individual columns. This is due to the fact `dag.nodes_in_topological_order()` sorts by qargs. I think this could be fixed, but it is a separate issue and affects all the drawers.

Sure, we can fix that in an a follow-on PR.

maddy-tod commented:
Just spotted that the `_get_image_depth()` still said it could raise an error, but that was no longer the case so have removed the comment - should be good to go now!

mtreinish commented:
> Just spotted that the `_get_image_depth()` still said it could raise an error, but that was no longer the case so have removed the comment - should be good to go now!

Good catch
